stingray/gti.py

```diff
@@ -1152,51 +1266,210 @@ def gti_border_bins(gtis, time, dt=None, epsilon=0.001):
     --------
     >>> times = np.arange(0.5, 13.5)
 
+    >>> gti_border_bins([[16., 18.]], times)
+    Traceback (most recent call last):
+        ...
+    ValueError: Invalid time interval for the given GTIs
+
     >>> start_bins, stop_bins = gti_border_bins(
     ...    [[0, 5], [6, 8]], times)
 
     >>> np.allclose(start_bins, [0, 6])
     True
     >>> np.allclose(stop_bins, [5, 8])
     True
-    >>> np.allclose(times[start_bins[0]:stop_bins[0]], [ 0.5, 1.5, 2.5, 3.5, 4.5])
+    >>> np.allclose(times[start_bins[0]:stop_bins[0]], [0.5, 1.5, 2.5, 3.5, 4.5])
     True
     >>> np.allclose(times[start_bins[1]:stop_bins[1]], [6.5, 7.5])
     True
     """
+    time = np.asarray(time)
+    gtis = np.asarray(gtis)
     if dt is None:
         dt = np.median(np.diff(time))
 
     epsilon_times_dt = epsilon * dt
 
+    if time[-1] < np.min(gtis) or time[0] > np.max(gtis):
+        raise ValueError("Invalid time interval for the given GTIs")
+
     spectrum_start_bins = []
     spectrum_stop_bins = []
 
-    for g in gtis:
-        # Epsilons here resolve a numerical problem when time edges end up
-        # outside GTIs.
-        good = ((time - dt / 2 >= g[0] - epsilon_times_dt) &
-                (time + dt / 2 <= g[1] + epsilon_times_dt))
-        t_good = time[good]
-        if len(t_good) == 0:
-            continue
-        startbin = np.argmin(np.abs(time - dt / 2 - g[0]))
-        stopbin = np.searchsorted(time + dt / 2, g[1], 'right') + 1
-        if stopbin > len(time):
-            stopbin = len(time)
+    gti_low = gtis[:, 0] + dt / 2 - epsilon_times_dt
+    gti_up = gtis[:, 1] - dt / 2 + epsilon_times_dt
+
+    for g0, g1 in zip(gti_low, gti_up):
+        startbin, stopbin = np.searchsorted(time, [g0, g1], "left")
+        stopbin += 1
+        if stopbin > time.size:
+            stopbin = time.size
 
-        if time[startbin] < (g[0] + dt / 2 - epsilon_times_dt):
+        if time[startbin] < g0:
             startbin += 1
         # Would be g[1] - dt/2, but stopbin is the end of an interval
         # so one has to add one bin
-        if time[stopbin - 1] > (g[1] - dt / 2 + epsilon_times_dt):
+        if time[stopbin - 1] > g1:
             stopbin -= 1
-        spectrum_start_bins.append([startbin])
-        spectrum_stop_bins.append([stopbin])
 
-    spectrum_start_bins = np.concatenate(spectrum_start_bins)
-    spectrum_stop_bins = np.concatenate(spectrum_stop_bins)
+        spectrum_start_bins.append(startbin)
+        spectrum_stop_bins.append(stopbin)
 
     assert len(spectrum_start_bins) > 0, \
         ("No GTIs are equal to or longer than chunk_length.")
-    return spectrum_start_bins, spectrum_stop_bins
+    return np.array(spectrum_start_bins), np.array(spectrum_stop_bins)
+
+
+def get_gti_events_idx(times, gti, dt=0):
+    """Get the indices of events from different GTIs of the observation.
```

Comment:
I ... am really struggling to understand the differences between the last four functions. All seem to have something to do with GTIs, and intervals, and finding the start and stop of them, but they all appear *really* similar somehow. :-/

Reply:
Hope the new names help :)